transaction.go

```diff
@@ -49,198 +90,303 @@ type Transaction struct {
-			}
-			pages[i].payload = make([]byte, payloadsize)
-			copy(pages[i].payload, data[i*maxPayloadSize:])
+	return buffer.Bytes(), nil
+}
+
+// unmarshalUpdates unmarshals the updates of a transaction
+func unmarshalUpdates(data []byte) ([]Update, error) {
+	buffer := bytes.NewBuffer(data)
+	updates := make([]Update, 0)
+
+	for {
+		update := Update{}
+
+		// Unmarshal name
+		var nameLength uint64
+		err1 := binary.Read(buffer, binary.LittleEndian, &nameLength)
+		if err1 == io.EOF {
+			// End of buffer reached
+			break
 		}
 
-		// Set the transaction number of the last page and increase the transactionCounter of the wal
-		pages[len(pages)-1].transactionNumber = t.wal.transactionCounter
-		t.wal.transactionCounter++
+		name := make([]byte, nameLength)
+		_, err2 := buffer.Read(name)
 
-		// Set the first and final page of the transaction
-		t.firstPage = &pages[0]
-		t.finalPage = &pages[len(pages)-1]
+		// Unmarshal version
+		var versionLength uint64
+		err3 := binary.Read(buffer, binary.LittleEndian, &versionLength)
 
-		// write the pages to disk and set the pageStatus
-		page := t.firstPage
-		for page != nil {
-			err := page.Write(t.wal.logFile)
-			if err != nil {
-				notifyChannel <- build.ExtendErr("Couldn't write the page to file", err)
-				return
-			}
-			page = page.nextPage
+		version := make([]byte, versionLength)
+		_, err4 := buffer.Read(version)
+
+		// Unmarshal instructions
+		var instructionsLength uint64
+		err5 := binary.Read(buffer, binary.LittleEndian, &instructionsLength)
+
+		instructions := make([]byte, instructionsLength)
+		_, err6 := buffer.Read(instructions)
+
+		// Check if any errors occured
+		if err1 != nil || err2 != nil || err3 != nil || err4 != nil || err5 != nil || err6 != nil {
+			return nil, errors.Compose(errors.New("Failed to unmarshal updates"), err1, err2, err3, err4, err5, err6)
 		}
 
-		// Commit transaction
-		notifyChannel <- t.commit()
-	}()
-	return notifyChannel
+		update.Name = string(name)
+		update.Version = string(version)
+		update.Instructions = instructions
+
+		updates = append(updates, update)
+	}
+
+	return updates, nil
 }
 
-// MarshalJSON marshales a transaction.
-func (t Transaction) checksum() ([crypto.HashSize]byte, error) {
+// threadedInitTransaction reserves pages of the wal, marshalls the
+// transactions's updates into a payload and splits the payload equally among
+// the pages. Once finished those pages are written to disk and the transaction
+// is committed.
+func threadedInitTransaction(t *Transaction) {
+	defer close(t.initComplete)
+
+	// Marshal all the updates to get their total length on disk
+	data, err := marshalUpdates(t.Updates)
+	if err != nil {
+		t.initComplete <- build.ExtendErr("could not marshal update", err)
+		return
+	}
+
+	// Find out how many pages are needed for the update
+	requiredPages := uint64(float64(len(data))/float64(maxPayloadSize+8) + 1)
+
+	// Get the pages from the wal
+	reservedPages := t.wal.reservePages(requiredPages)
```

Comment:
The `reservePages` function needs to be called 'managedReservePages' because you are expected to call it without holding a lock.

The `reservePages` function should probably fill out most of the page fields for you, instead of you needing to do it after you reserve them. For example, the pageStatus on the first page should be automatically set to '2', and the pageStatus on all additional pages should be set to '1'. The page.offset fields can be already set to the correct offsets, and the nextPage pointers can also be connected together correctly.

If you pass in the payload directly instead of passing in the number of pages you need, it can fill out the payload as well, which would include figuring out the correct number of pages to provide.

That would clean up the `commit` function quite a bit I think.